persistence/jdbc/src/main/java/org/infinispan/persistence/jdbc/impl/table/AbstractTableManager.java

```diff
@@ -21,36 +21,62 @@
-   protected TableName tableName;
-
-   protected String insertRowSql;
-   protected String updateRowSql;
-   protected String upsertRowSql;
-   protected String selectRowSql;
-   protected String selectMultipleRowSql;
-   protected String selectIdRowSql;
-   protected String deleteRowSql;
-   protected String loadAllRowsSql;
-   protected String countRowsSql;
-   protected String loadAllNonExpiredRowsSql;
-   protected String deleteAllRows;
-   protected String selectExpiredRowsSql;
-   protected String deleteExpiredRowsSql;
-
-   AbstractTableManager(ConnectionFactory connectionFactory, TableManipulationConfiguration config, DbMetaData metaData, Log log) {
+   protected final String identifierQuoteString;
+   protected final DbMetaData metaData;
+   protected final TableName tableName;
+
+   protected final String insertRowSql;
+   protected final String updateRowSql;
+   protected final String upsertRowSql;
+   protected final String selectRowSql;
+   //protected final String selectMultipleRowSql;
+   protected final String selectIdRowSql;
+   protected final String deleteRowSql;
+   protected final String loadAllRowsSql;
+   protected final String countRowsSql;
+   protected final String loadAllNonExpiredRowsSql;
+   //protected final String deleteAllRows;
+   protected final String selectExpiredRowsSql;
+   protected final String deleteExpiredRowsSql;
+
+   AbstractTableManager(ConnectionFactory connectionFactory, TableManipulationConfiguration config, DbMetaData metaData, String cacheName, Log log) {
+      this(connectionFactory, config, metaData, cacheName, DEFAULT_IDENTIFIER_QUOTE_STRING, log);
+   }
+
+   AbstractTableManager(ConnectionFactory connectionFactory, TableManipulationConfiguration config, DbMetaData metaData, String cacheName, String identifierQuoteString, Log log) {
+      // cacheName is required
+      if (cacheName == null || cacheName.trim().length() == 0)
+         throw new PersistenceException("cacheName needed in order to create table");
+
       this.connectionFactory = connectionFactory;
       this.config = config;
       this.metaData = metaData;
+      this.tableName = new TableName(identifierQuoteString, config.tableNamePrefix(), cacheName);
+      this.identifierQuoteString = identifierQuoteString;
       this.log = log;
+
+      // init row sql
+      this.insertRowSql = getInsertRowSql();
+      this.updateRowSql = getUpdateRowSql();
+      this.upsertRowSql = getUpsertRowSql();
+      this.selectRowSql = getSelectRowSql();
+      // selectMultipleRowSql
+      this.selectIdRowSql = getSelectIdRowSql();
+      this.deleteRowSql = getDeleteRowSql();
+      this.loadAllRowsSql = getLoadAllRowsSql();
+      this.countRowsSql = getCountNonExpiredRowsSql();
+      this.loadAllNonExpiredRowsSql = getLoadNonExpiredAllRowsSql();
+      // deleteAllRows
+      this.selectExpiredRowsSql = getSelectExpiredBucketsSql();
+      this.deleteExpiredRowsSql = getSelectOnlyExpiredRowsSql();
```

Comment:
None of these fields are ever used once you have initialised them, as the store always calls the getter methods which at the moment creates a String on every invocation. This is what the null check in the previous implementation was used to avoid, basically it's purpose was to only init the variable once in a lazy manner.

I think populating the variables, e.g. `insertRowSql`, upfront and making them final is desirable ... but you will have to change your approach. I suggest that `getInsertRowSQL` just returns the `insertRowSql` variable and in the constructor you call a protected method `initInsertRowSQL` to init the variable. That way, concrete implementations of the AbstractTableManager can still overload the init method where required and it will allow the holding variables to be `private final`.

Reply:
Also, in these init methods it's probably best to just reference the `tableName` variable directly as there is no benefit of calling the method `getTableName()`.

Reply:
Yeah. You are right. I didn't pay attention to that.

Looking with more details now I saw the following issue:

```java
public class FooTest {

   private final String foo;
   private final String bar;

   public FooTest() {
      foo = initFoo();
      bar = initBar();
   }

   private String initFoo() {
      return "foo" + bar;
   }

   private String initBar() {
      return "bar";
   }
}
```

Using the `initSql` approach we will be facing `null`values. See https://github.com/infinispan/infinispan/blob/ce493e14b30da7d08ffb24cfb43b78687c769d8a/persistence/jdbc/src/main/java/org/infinispan/persistence/jdbc/impl/table/MySQLTableManager.java#L29 The method `getInsertRowSql()` is being called from `getUpsertRowSql()`. There are more entries like the example above.

That issue could be fixed ordering the variables in the constructor but I would like to hear from you what do you think.

Reply:
I think ordering the variables appropriately is fine for now.